# Estructuras Dataframe

In [1]:
import pandas as pd

In [2]:
vuelos=pd.read_csv('DelayedFlights.csv')
vuelos

Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
0         2211.0        2225            WN  ...     4.0     8.0          0   
1         1002.0        1000            WN  ...     5.0    10.0          0   
2          804.0         750            WN  ...     3.0    17.0          0   
3         1959.0        1925            WN  ...     3.0    10.0          0   
4         2121.0        2110            WN  ...     4.0    10.0          0   
...          ...         ...           ...  ...     ...     ...        ...   
1936753   1617.0        1552            DL  ...     9.0    18.0          0   
1936754    904.0         749            DL  ...    15.0    34.0          0   
1936755   1149.0        1010            DL  ...     8.0    32.0          0   
1936756   1446.0        1437            DL  ...    13.0    13.0          0   
1936757   1413.0        1418            DL  ...     8.0    11.0          0   

         CancellationCode  Diverted  CarrierDelay  WeatherDelay NASDelay  \
0                       N         0           NaN           NaN      NaN   
1                       N         0           NaN           NaN      NaN   
2                       N         0           NaN           NaN      NaN   
3                       N         0           2.0           0.0      0.0   
4                       N         0           NaN           NaN      NaN   
...                   ...       ...           ...           ...      ...   
1936753                 N         0           3.0           0.0      0.0   
1936754                 N         0           0.0          57.0     18.0   
1936755                 N         0           1.0           0.0     19.0   
1936756                 N         0           NaN           NaN      NaN   
1936757                 N         0           NaN           NaN      NaN   

        SecurityDelay  LateAircraftDelay  
0                 NaN                NaN  
1                 NaN                NaN  
2                 NaN                NaN  
3                 0.0               32.0  
4                 NaN                NaN  
...               ...                ...  
1936753           0.0               22.0  
1936754           0.0                0.0  
1936755           0.0               79.0  
1936756           NaN                NaN  
1936757           NaN                NaN  

[1936758 rows x 30 columns]

In [3]:
#Reducción a 200000 registros (filas) para trabajar en todos los ejercicios

vuelos=vuelos.sample(200000)

# Ejercicio 1. 
## Visualización de todas las variables existentes en el DataFrame para la elección de las que tengan información relevante para su análisis estadístico, justificándolo.

In [4]:
# Comprobación de las variables (o columnas) existentes.

vuelos.columns.tolist()

['Unnamed: 0',
 'Year',
 'Month',
 'DayofMonth',
 'DayOfWeek',
 'DepTime',
 'CRSDepTime',
 'ArrTime',
 'CRSArrTime',
 'UniqueCarrier',
 'FlightNum',
 'TailNum',
 'ActualElapsedTime',
 'CRSElapsedTime',
 'AirTime',
 'ArrDelay',
 'DepDelay',
 'Origin',
 'Dest',
 'Distance',
 'TaxiIn',
 'TaxiOut',
 'Cancelled',
 'CancellationCode',
 'Diverted',
 'CarrierDelay',
 'WeatherDelay',
 'NASDelay',
 'SecurityDelay',
 'LateAircraftDelay']

### Información sobre el significado de cada variable

"Year":
año del vuelo (todos del 2008).

"Month":
mes del vuelo.

"DayOfMonth":
día del mes (1 a 31).

"DayOfWeek":
día de la semana.

"DepTime":
hora de salida real.

"CRSDepTime":
hora de salida programada.

"ArrTime":
hora de llegada real.

"CRSArrTime":
hora de llegada programada.

"UniqueCarrier":
ID de operador.

"FlightNum":
número de vuelo.

"TailNum":
número de cola del avión.

"ActualElapsedTime":
tiempo real transcurrido del vuelo, en minutos.

"CRSElapsedTime":
tiempo transcurrido programado del vuelo, en minutos.

"AirTime":
tiempo en el aire para el vuelo, en minutos.

"ArrDelay":
retardo de llegada, en minutos.

"DepDelay":
retraso de salida, en minutos.

"Origin":
aeropuerto de origen.

"Dest":
aeropuerto de destino.

"Distance":
distancia de vuelo, en millas.

"TaxiIn":
tiempo de rodaje desde las ruedas hasta la llegada a la puerta, en minutos.

"TaxiOut":
tiempo de rodaje desde la salida de la puerta hasta las ruedas, en minutos.

"Cancelled":
estado de cancelación, donde Sí=1 y No=0.

"CancellationCode":
código de cancelación, en caso de existir.

"Diverted":
estado de desvío, donde Sí=1 y No=0.

"CarrierDelay":
retardo, en minutos, atribuible al transportista.

"WeatherDelay":
retraso, en minutos, atribuible a factores meteorológicos.

"NASDelay":
retraso, en minutos, atribuible al Sistema Nacional de Aviación.

"SecurityDelay":
retardo, en minutos, atribuible a factores de seguridad.

"LateAircraftDelay":
retraso, en minutos, atribuible a aeronaves que llegan tarde.

### Eliminación de columnas

Como el objetivo del nuevo DataFrame es su análisis estadístico sobre los tiempos de los vuelos, he eliminado las variables (o columnas) que carecen de tal fin, tales como:

'Unnamed: 0', aparentemente tiene la forma de un indice para el dataset, aunque no tiene un orden consecutivo. Aún así, pandas ya generará un indice ordenado de números para cada registro; 

'Year', ya que todos los registros corresponden al año 2008; 

'FlightNUm' y 'TailNum', ambas solo contienen información de interés al ámbito de control interno para las compañías aéreas;

'TaxiIn' y 'TaxiOut', ya están dentro del tiempo de vuelo;

'CancellationCode', lo que interesa es disponer de si existe cancelación o no del vuelo (algo con lo que el dataframe ya cuenta);

'CarrierDelay', 'WeatherDelay', 'NASDelay', ' 'SecurityDelay' y 'LateAircraftDelay', aquí no analizamos la responsabilidad por el retraso del vuelo, sino los tiempos que se han empleado (además, muchos de sus valores carecen de registro).

### En consecuencia, el nuevo dataframe quedaría de la siguiente manera:

In [5]:
vuelos.drop(['Unnamed: 0', 'Year', 'FlightNum', 'TailNum', 'TaxiIn', 'TaxiOut', 'CancellationCode', 
             'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay'], axis=1, inplace=True)
vuelos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
505472       3           8          6   1520.0        1455   1619.0   
1062578      6          14          6   2252.0        2005    311.0   
1736297     12           6          6   1359.0        1350   1634.0   
132041       1          21          1   2259.0        2140    600.0   
1622652     10          11          6   1534.0        1525   1754.0   
...        ...         ...        ...      ...         ...      ...   
1607025     10          28          2   2232.0        2105    114.0   
856168       5          28          3   1707.0        1655   2253.0   
1652900     11          15          6   1441.0        1229   1600.0   
1814438     12          21          7    944.0         842   1107.0   
1744271     12          16          2   1956.0        1940   2104.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
505472         1600            MQ               59.0            65.0     38.0   
1062578        2237            B6              379.0           272.0    220.0   
1736297        1650            WN              275.0           300.0    261.0   
132041          505            NW              241.0           265.0    216.0   
1622652        1803            CO              260.0           278.0    236.0   
...             ...           ...                ...             ...      ...   
1607025        2359            AA              162.0           174.0    139.0   
856168         2230            AA              226.0           215.0    204.0   
1652900        1356            YV               79.0            87.0     61.0   
1814438        1014            UA              143.0           152.0    128.0   
1744271        2055            WN               68.0            75.0     54.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
505472       19.0      25.0    LSE  ORD       215          0         0  
1062578     274.0     167.0    JFK  DEN      1626          0         0  
1736297     -16.0       9.0    RDU  PHX      1891          0         0  
132041       55.0      79.0    SFO  DTW      2079          0         0  
1622652      -9.0       9.0    IAH  SEA      1874          0         0  
...           ...       ...    ...  ...       ...        ...       ...  
1607025      75.0      87.0    MIA  JFK      1090          0         0  
856168       23.0      12.0    SEA  DFW      1660          0         0  
1652900     124.0     132.0    IAD  CLT       321          0         0  
1814438      53.0      62.0    DEN  SNA       846          0         0  
1744271       9.0      16.0    BUR  SMF       358          0         0  

[200000 rows x 18 columns]

## Ejercicio 2

In [6]:
# a) Resumen estadístico y análisis de datos relevantes (no habituales o evidentes)
vuelos[['CRSDepTime', 'ActualElapsedTime',
        'AirTime', 'DepDelay', 'ArrDelay', 'Distance']].describe().round(2)

CRSDepTime  ActualElapsedTime    AirTime   DepDelay   ArrDelay  \
count   200000.00          199181.00  199181.00  200000.00  199181.00   
mean      1467.86             133.39     108.36      43.40      42.40   
std        425.63              71.99      68.56      53.42      56.83   
min          1.00              16.00       0.00       6.00     -68.00   
25%       1133.00              80.00      58.00      12.00       9.00   
50%       1510.00             116.00      90.00      24.00      25.00   
75%       1819.00             165.00     137.00      54.00      56.00   
max       2359.00             727.00     637.00    1490.00    1490.00   

        Distance  
count  200000.00  
mean      766.59  
std       574.07  
min        11.00  
25%       338.00  
50%       607.00  
75%      1005.00  
max      4962.00

Como puede observarse, la mitad de los vuelos se efectúan pasado poco más el mediodía (unos minutos después de las 15:00), y solo tres horas más tarde (alrededor de las 18:00) ya se han realizado el 75%.

Por otra parte, la media del transcurso real de los vuelos es de 133.39 minutos.

También llama la atención el vuelo que emplea solo un minuto en el aire (* en la siguiente línea podrá verse tal/es registro/s), Como es lógico, se trata de algún fallo en el mismo, mas no de un 'outlier', ya que el la distancia de estos vuelos llega a las 932 millas.

El mismo tiempo de retraso de las salidas se traslada al de las llegadas, así se puede ver en sus respectivas medias, las cuales son prácticamente idénticas.

Los datos registrados en los retrasos de llegada son verdaderamente extremos, hay mucha dispersión. Así lo reflejan sus desviaciones típicas, las cuales están incluso por encima de sus respectivas medias. Destacando en los retrasos de llegada un valor en negativo, lo que supone un adelanto en la llegada con respecto a la hora programada de 68 minutos (**).

Por último, cabe resaltar la distancia máxima de 4962 kms y la mínima de 30 kms, también con una desviación típida muy extensa, y con una media de 766.59 kms recorridos por cada vuelo.

In [7]:
# * visualización del resgistro con valor de 1 minuto en el tiempo de vuelo
vuelos[vuelos.AirTime==1]


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1178266      7          11          5   1015.0        1002   1249.0   
950529       6          17          2   1136.0        1130   1213.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
1178266        1234            EV              154.0           152.0      1.0   
950529         1211            OO               37.0            41.0      1.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
1178266      15.0      13.0    FLL  CVG       932          0         0  
950529        2.0       6.0    LAX  OXR        49          0         0

In [28]:
# ** visualización del registro con máximo adelanto de la llegada
vuelos[vuelos.ArrDelay==-68]

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
352034      2          13          3    907.0         900   1202.0   

        CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
352034        1310            AQ              295.0           370.0  ...   

        ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
352034     -68.0       7.0     SNA  KOA     2530          0         0   

        AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
352034        534.51        WithinTime               11.0  

[1 rows x 21 columns]

### b) Datos faltantes por columna


In [9]:
vuelos.isna().sum()


Month                  0
DayofMonth             0
DayOfWeek              0
DepTime                0
CRSDepTime             0
ArrTime              707
CRSArrTime             0
UniqueCarrier          0
ActualElapsedTime    819
CRSElapsedTime        19
AirTime              819
ArrDelay             819
DepDelay               0
Origin                 0
Dest                   0
Distance               0
Cancelled              0
Diverted               0
dtype: int64

In [10]:
print('Si sumamos todos los datos faltantes, hay un total de', vuelos.isna().sum().sum(), 'valores carentes de contenido.')

Si sumamos todos los datos faltantes, hay un total de 3183 valores carentes de contenido.


In [11]:
#Comprobación de los registros (o filas) con valores nulos

vuelos[vuelos.isna().any(axis=1)]


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
1747420     12          17          3   2054.0        1940      NaN   
275915       2          26          2   1326.0        1311      NaN   
1918040     12          19          5   1038.0         940      NaN   
312727       2           4          1   1347.0        1305      NaN   
1909951     12          26          5   1259.0        1245   2151.0   
...        ...         ...        ...      ...         ...      ...   
319456       2           4          1   1806.0        1720      NaN   
1809146     12          22          1    934.0         922      NaN   
1765191     12          26          5   1332.0        1240      NaN   
116475       1           6          7   1427.0        1420      NaN   
1752082     12          19          5   1501.0        1245    225.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
1747420        2055            WN                NaN            75.0      NaN   
275915         1907            US                NaN           236.0      NaN   
1918040        1117            B6                NaN            97.0      NaN   
312727         1407            MQ                NaN            62.0      NaN   
1909951        1850            AA                NaN           245.0      NaN   
...             ...           ...                ...             ...      ...   
319456           34            NW                NaN           254.0      NaN   
1809146        1037            OO                NaN           135.0      NaN   
1765191        1700            WN                NaN           200.0      NaN   
116475         1516            MQ                NaN            56.0      NaN   
1752082        1525            WN                NaN           100.0      NaN   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  
1747420       NaN      74.0    SAN  LAS       258          0         1  
275915        NaN      15.0    PHX  TPA      1788          0         1  
1918040       NaN      58.0    BOS  BUF       396          0         1  
312727        NaN      42.0    DBQ  ORD       147          0         1  
1909951       NaN      14.0    EGE  MIA      1811          0         1  
...           ...       ...    ...  ...       ...        ...       ...  
319456        NaN      46.0    LAX  DTW      1979          0         1  
1809146       NaN      12.0    SYR  ORD       607          1         0  
1765191       NaN      52.0    PHX  MDW      1444          0         1  
116475        NaN       7.0    SJC  SAN       417          0         1  
1752082       NaN     136.0    PDX  SLC       630          0         1  

[819 rows x 18 columns]

In [12]:
vuelos.isna().any(axis=1).sum()

819

Como la cantidad de las filas con datos nulos son prácticamente insignificantes en proporción a la cantidad de filas totales (no llega ni al 0.5%), y con la finalidad de mantener todos los registros, la cancelación de estos valores no inlfuirá en análisis de cualquier tipo que se realice sobre el dataset.

#### Eliminación de las filas con datos nulos

In [13]:
c=vuelos.apply(lambda x: x['Cancelled']==1, axis=1).sum()
d=vuelos.apply(lambda x: x['Diverted']==1, axis=1).sum()

print(f'Hay {c} vuelos canceldos')

print(f'Hay {d} vuelos desviados')

print('Puede observarse que los valores nulos pertenecen a los registros que constan como vuelos nulos y cancelados:', c+d,',\ncoincidentes con los datos nulos totales de la variables ActualElapsedTime, AirTime y ArrDelay')


Hay 67 vuelos canceldos
Hay 752 vuelos desviados
Puede observarse que los valores nulos pertenecen a los registros que constan como vuelos nulos y cancelados: 819 ,
coincidentes con los datos nulos totales de la variables ActualElapsedTime, AirTime y ArrDelay


In [14]:
# Eliminación de los datos nulos después de la información señalada

vuelos = vuelos.drop(vuelos[vuelos['Cancelled']==1].index)
vuelos = vuelos.drop(vuelos[vuelos['Diverted']==1].index)
vuelos.isna().sum()




Month                0
DayofMonth           0
DayOfWeek            0
DepTime              0
CRSDepTime           0
ArrTime              0
CRSArrTime           0
UniqueCarrier        0
ActualElapsedTime    0
CRSElapsedTime       0
AirTime              0
ArrDelay             0
DepDelay             0
Origin               0
Dest                 0
Distance             0
Cancelled            0
Diverted             0
dtype: int64

### c) Creación de nuevas columnas

#### Velocidad media por vuelo

In [15]:
vuelos['AvarageSpeed']= (vuelos['Distance'] / vuelos['AirTime']*60).round(2)
vuelos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
505472       3           8          6   1520.0        1455   1619.0   
1062578      6          14          6   2252.0        2005    311.0   
1736297     12           6          6   1359.0        1350   1634.0   
132041       1          21          1   2259.0        2140    600.0   
1622652     10          11          6   1534.0        1525   1754.0   
...        ...         ...        ...      ...         ...      ...   
1607025     10          28          2   2232.0        2105    114.0   
856168       5          28          3   1707.0        1655   2253.0   
1652900     11          15          6   1441.0        1229   1600.0   
1814438     12          21          7    944.0         842   1107.0   
1744271     12          16          2   1956.0        1940   2104.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
505472         1600            MQ               59.0            65.0     38.0   
1062578        2237            B6              379.0           272.0    220.0   
1736297        1650            WN              275.0           300.0    261.0   
132041          505            NW              241.0           265.0    216.0   
1622652        1803            CO              260.0           278.0    236.0   
...             ...           ...                ...             ...      ...   
1607025        2359            AA              162.0           174.0    139.0   
856168         2230            AA              226.0           215.0    204.0   
1652900        1356            YV               79.0            87.0     61.0   
1814438        1014            UA              143.0           152.0    128.0   
1744271        2055            WN               68.0            75.0     54.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  \
505472       19.0      25.0    LSE  ORD       215          0         0   
1062578     274.0     167.0    JFK  DEN      1626          0         0   
1736297     -16.0       9.0    RDU  PHX      1891          0         0   
132041       55.0      79.0    SFO  DTW      2079          0         0   
1622652      -9.0       9.0    IAH  SEA      1874          0         0   
...           ...       ...    ...  ...       ...        ...       ...   
1607025      75.0      87.0    MIA  JFK      1090          0         0   
856168       23.0      12.0    SEA  DFW      1660          0         0   
1652900     124.0     132.0    IAD  CLT       321          0         0   
1814438      53.0      62.0    DEN  SNA       846          0         0   
1744271       9.0      16.0    BUR  SMF       358          0         0   

         AvarageSpeed  
505472         339.47  
1062578        443.45  
1736297        434.71  
132041         577.50  
1622652        476.44  
...               ...  
1607025        470.50  
856168         488.24  
1652900        315.74  
1814438        396.56  
1744271        397.78  

[199181 rows x 19 columns]

#### Vuelo con retraso o dentro de hora

In [16]:
import numpy as np

vuelos['Delay_WithinTime']=np.where(vuelos['ArrDelay']>0, 'Delay', 'WithinTime')
vuelos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
505472       3           8          6   1520.0        1455   1619.0   
1062578      6          14          6   2252.0        2005    311.0   
1736297     12           6          6   1359.0        1350   1634.0   
132041       1          21          1   2259.0        2140    600.0   
1622652     10          11          6   1534.0        1525   1754.0   
...        ...         ...        ...      ...         ...      ...   
1607025     10          28          2   2232.0        2105    114.0   
856168       5          28          3   1707.0        1655   2253.0   
1652900     11          15          6   1441.0        1229   1600.0   
1814438     12          21          7    944.0         842   1107.0   
1744271     12          16          2   1956.0        1940   2104.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  AirTime  \
505472         1600            MQ               59.0            65.0     38.0   
1062578        2237            B6              379.0           272.0    220.0   
1736297        1650            WN              275.0           300.0    261.0   
132041          505            NW              241.0           265.0    216.0   
1622652        1803            CO              260.0           278.0    236.0   
...             ...           ...                ...             ...      ...   
1607025        2359            AA              162.0           174.0    139.0   
856168         2230            AA              226.0           215.0    204.0   
1652900        1356            YV               79.0            87.0     61.0   
1814438        1014            UA              143.0           152.0    128.0   
1744271        2055            WN               68.0            75.0     54.0   

         ArrDelay  DepDelay Origin Dest  Distance  Cancelled  Diverted  \
505472       19.0      25.0    LSE  ORD       215          0         0   
1062578     274.0     167.0    JFK  DEN      1626          0         0   
1736297     -16.0       9.0    RDU  PHX      1891          0         0   
132041       55.0      79.0    SFO  DTW      2079          0         0   
1622652      -9.0       9.0    IAH  SEA      1874          0         0   
...           ...       ...    ...  ...       ...        ...       ...   
1607025      75.0      87.0    MIA  JFK      1090          0         0   
856168       23.0      12.0    SEA  DFW      1660          0         0   
1652900     124.0     132.0    IAD  CLT       321          0         0   
1814438      53.0      62.0    DEN  SNA       846          0         0   
1744271       9.0      16.0    BUR  SMF       358          0         0   

         AvarageSpeed Delay_WithinTime  
505472         339.47            Delay  
1062578        443.45            Delay  
1736297        434.71       WithinTime  
132041         577.50            Delay  
1622652        476.44       WithinTime  
...               ...              ...  
1607025        470.50            Delay  
856168         488.24            Delay  
1652900        315.74            Delay  
1814438        396.56            Delay  
1744271        397.78            Delay  

[199181 rows x 20 columns]

#### (Creación de nueva columna) Tiempo de carreteo o movimiento del avión sobre tierra

In [17]:

vuelos['TaxiIn_and_TaxiOut']=(vuelos['ActualElapsedTime'] - vuelos['AirTime']).round(2)
vuelos


Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
505472       3           8          6   1520.0        1455   1619.0   
1062578      6          14          6   2252.0        2005    311.0   
1736297     12           6          6   1359.0        1350   1634.0   
132041       1          21          1   2259.0        2140    600.0   
1622652     10          11          6   1534.0        1525   1754.0   
...        ...         ...        ...      ...         ...      ...   
1607025     10          28          2   2232.0        2105    114.0   
856168       5          28          3   1707.0        1655   2253.0   
1652900     11          15          6   1441.0        1229   1600.0   
1814438     12          21          7    944.0         842   1107.0   
1744271     12          16          2   1956.0        1940   2104.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
505472         1600            MQ               59.0            65.0  ...   
1062578        2237            B6              379.0           272.0  ...   
1736297        1650            WN              275.0           300.0  ...   
132041          505            NW              241.0           265.0  ...   
1622652        1803            CO              260.0           278.0  ...   
...             ...           ...                ...             ...  ...   
1607025        2359            AA              162.0           174.0  ...   
856168         2230            AA              226.0           215.0  ...   
1652900        1356            YV               79.0            87.0  ...   
1814438        1014            UA              143.0           152.0  ...   
1744271        2055            WN               68.0            75.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
505472       19.0      25.0     LSE  ORD      215          0         0   
1062578     274.0     167.0     JFK  DEN     1626          0         0   
1736297     -16.0       9.0     RDU  PHX     1891          0         0   
132041       55.0      79.0     SFO  DTW     2079          0         0   
1622652      -9.0       9.0     IAH  SEA     1874          0         0   
...           ...       ...     ...  ...      ...        ...       ...   
1607025      75.0      87.0     MIA  JFK     1090          0         0   
856168       23.0      12.0     SEA  DFW     1660          0         0   
1652900     124.0     132.0     IAD  CLT      321          0         0   
1814438      53.0      62.0     DEN  SNA      846          0         0   
1744271       9.0      16.0     BUR  SMF      358          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
505472         339.47             Delay               21.0  
1062578        443.45             Delay              159.0  
1736297        434.71        WithinTime               14.0  
132041         577.50             Delay               25.0  
1622652        476.44        WithinTime               24.0  
...               ...               ...                ...  
1607025        470.50             Delay               23.0  
856168         488.24             Delay               22.0  
1652900        315.74             Delay               18.0  
1814438        396.56             Delay               15.0  
1744271        397.78             Delay               14.0  

[199181 rows x 21 columns]

### d) Tabla de aerolíneas con más retrasos acumulados. Muestra de mayor a menor

In [18]:
#Creación de un dataframe solo con valores 'Delay' (retraso) de la columna ya creada anteriormente 'Delay_WithinTime'

solo_retrasos=vuelos.groupby('Delay_WithinTime').get_group('Delay')
solo_retrasos

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
505472       3           8          6   1520.0        1455   1619.0   
1062578      6          14          6   2252.0        2005    311.0   
132041       1          21          1   2259.0        2140    600.0   
1536998     10          20          1    712.0         705    924.0   
868298       5          19          1   1931.0        1805   2209.0   
...        ...         ...        ...      ...         ...      ...   
1607025     10          28          2   2232.0        2105    114.0   
856168       5          28          3   1707.0        1655   2253.0   
1652900     11          15          6   1441.0        1229   1600.0   
1814438     12          21          7    944.0         842   1107.0   
1744271     12          16          2   1956.0        1940   2104.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
505472         1600            MQ               59.0            65.0  ...   
1062578        2237            B6              379.0           272.0  ...   
132041          505            NW              241.0           265.0  ...   
1536998         915            WN              132.0           130.0  ...   
868298         2103            B6              158.0           178.0  ...   
...             ...           ...                ...             ...  ...   
1607025        2359            AA              162.0           174.0  ...   
856168         2230            AA              226.0           215.0  ...   
1652900        1356            YV               79.0            87.0  ...   
1814438        1014            UA              143.0           152.0  ...   
1744271        2055            WN               68.0            75.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
505472       19.0      25.0     LSE  ORD      215          0         0   
1062578     274.0     167.0     JFK  DEN     1626          0         0   
132041       55.0      79.0     SFO  DTW     2079          0         0   
1536998       9.0       7.0     TPA  BWI      842          0         0   
868298       66.0      86.0     FLL  LGA     1076          0         0   
...           ...       ...     ...  ...      ...        ...       ...   
1607025      75.0      87.0     MIA  JFK     1090          0         0   
856168       23.0      12.0     SEA  DFW     1660          0         0   
1652900     124.0     132.0     IAD  CLT      321          0         0   
1814438      53.0      62.0     DEN  SNA      846          0         0   
1744271       9.0      16.0     BUR  SMF      358          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
505472         339.47             Delay               21.0  
1062578        443.45             Delay              159.0  
132041         577.50             Delay               25.0  
1536998        417.52             Delay               11.0  
868298         454.65             Delay               16.0  
...               ...               ...                ...  
1607025        470.50             Delay               23.0  
856168         488.24             Delay               22.0  
1652900        315.74             Delay               18.0  
1814438        396.56             Delay               15.0  
1744271        397.78             Delay               14.0  

[178059 rows x 21 columns]

In [19]:
# Creación de un dataframe solo con valores 'Delay' (retraso) de la columna ya creada anteriormente 'Delay_WithinTime'
solo_retrasos=vuelos.groupby('Delay_WithinTime').get_group('Delay')


# Contabilización vuelos con retrasos por aerolínea
retrasos_aerol=pd.DataFrame(solo_retrasos.groupby('UniqueCarrier')['Delay_WithinTime'].value_counts())


# Contabilización vuelos totales por aerolínea
total_vuelos_aerol=pd.DataFrame(vuelos.groupby('UniqueCarrier')['Delay_WithinTime'].count())

# Porcentaje de vuelos retrasados con respecto a los vuelos totales por aerolínea
porcent_retr=(retrasos_aerol/total_vuelos_aerol)*100

porcent_retr.columns=['Porcentaje']
porcent_retr.sort_values(by='Porcentaje', ascending=False)



Porcentaje
UniqueCarrier Delay_WithinTime            
HA            Delay              95.572917
YV            Delay              94.727011
OH            Delay              93.497592
OO            Delay              92.560223
MQ            Delay              92.525702
NW            Delay              91.787026
EV            Delay              91.576443
XE            Delay              91.510227
FL            Delay              91.475635
F9            Delay              90.699301
9E            Delay              90.581351
AA            Delay              90.453715
DL            Delay              88.879585
UA            Delay              88.503269
B6            Delay              87.445737
AS            Delay              87.229713
WN            Delay              86.403071
US            Delay              85.054750
AQ            Delay              84.848485
CO            Delay              83.711021

### e)  Los vuelos más largos. Muestra de los 10 primeros para tener una buena visión de esta variable, pero sin llegar a ser excesiva. 
#### # Matizar que el ejercicio no pide una nueva tabla solo con los valores de la columna de 'Distancia' o alguna más, de lo contrario habría procedido como en el ejercicio anterior. Se entiende que presente los registros de los vuelos más largos.

In [20]:
vuelos.sort_values(by='Distance', ascending=False).head(10)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
719889       4          20          7   2316.0        1955   1510.0   
1415598      8          13          3   1344.0        1335   1814.0   
1926347     12          20          6   1441.0        1335   2123.0   
1068315      6           2          1   1342.0        1335   1807.0   
1725876     11          22          6   1413.0        1335   1958.0   
718540       4          30          3   1408.0        1335   1843.0   
556358       3          18          2   2207.0        2145   1326.0   
556366       3           5          3   1335.0        1320   1920.0   
1421428      8          22          5   1344.0        1335   1815.0   
176119       1          26          6   1407.0        1320   1947.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
719889         1130            CO              594.0           575.0  ...   
1415598        1810            CO              630.0           635.0  ...   
1926347        1935            CO              702.0           660.0  ...   
1068315        1810            CO              625.0           635.0  ...   
1725876        1935            CO              645.0           660.0  ...   
718540         1810            CO              635.0           635.0  ...   
556358         1310            CO              559.0           565.0  ...   
556366         1920            CO              645.0           660.0  ...   
1421428        1810            CO              631.0           635.0  ...   
176119         1920            CO              640.0           660.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
719889      220.0     201.0     HNL  EWR     4962          0         0   
1415598       4.0       9.0     EWR  HNL     4962          0         0   
1926347     108.0      66.0     EWR  HNL     4962          0         0   
1068315      -3.0       7.0     EWR  HNL     4962          0         0   
1725876      23.0      38.0     EWR  HNL     4962          0         0   
718540       33.0      33.0     EWR  HNL     4962          0         0   
556358       16.0      22.0     HNL  EWR     4962          0         0   
556366        0.0      15.0     EWR  HNL     4962          0         0   
1421428       5.0       9.0     EWR  HNL     4962          0         0   
176119       27.0      47.0     EWR  HNL     4962          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
719889         537.40             Delay               40.0  
1415598        497.86             Delay               32.0  
1926347        468.85             Delay               67.0  
1068315        502.06        WithinTime               32.0  
1725876        488.87             Delay               36.0  
718540         484.10             Delay               20.0  
556358         570.34             Delay               37.0  
556366         478.65        WithinTime               23.0  
1421428        490.48             Delay               24.0  
176119         491.29             Delay               34.0  

[10 rows x 21 columns]

In [21]:
pd.DataFrame(vuelos.groupby('UniqueCarrier')['Delay_WithinTime'].count())

Delay_WithinTime
UniqueCarrier                  
9E                         5298
AA                        19704
AQ                           66
AS                         3931
B6                         5759
CO                        10326
DL                        11942
EV                         8405
F9                         2860
FL                         7285
HA                          768
MQ                        14396
NW                         8170
OH                         5398
OO                        13616
UA                        14378
US                        10137
WN                        39075
XE                        10707
YV                         6960

### Los 10 vuelos con más retrasos. 
#### Procedimiento idem ejercicio anterior. 

In [22]:
vuelos.sort_values(by='ArrDelay', ascending=False).head(10)

Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  ArrTime  \
521096       3          21          5    705.0         615    811.0   
1222506      7          30          3   1241.0        1445   2110.0   
1400101      8          29          5    630.0         850    956.0   
457238       3          29          6   1013.0        1305   1322.0   
325724       2           3          7    939.0        1400   1358.0   
1906664     12          17          3   1103.0        1430   1555.0   
1599350     10          30          4    711.0        1205   1244.0   
315641       2          29          5    713.0        1210    917.0   
1698976     11          30          7   1115.0        1655   1318.0   
1879242     12          21          7     27.0         615    222.0   

         CRSArrTime UniqueCarrier  ActualElapsedTime  CRSElapsedTime  ...  \
521096          721            NW               66.0            66.0  ...   
1222506        2305            AA              329.0           320.0  ...   
1400101        1230            AA              146.0           160.0  ...   
457238         1615            UA              369.0           370.0  ...   
325724         1709            NW              199.0           129.0  ...   
1906664        1930            AA              172.0           180.0  ...   
1599350        1736            NW              213.0           211.0  ...   
315641         1415            MQ              124.0           125.0  ...   
1698976        1855            MQ              123.0           120.0  ...   
1879242         759            MQ              115.0           104.0  ...   

         ArrDelay  DepDelay  Origin Dest Distance  Cancelled  Diverted  \
521096     1490.0    1490.0     BNA  MEM      200          0         0   
1222506    1325.0    1316.0     HNL  LAX     2556          0         0   
1400101    1286.0    1300.0     SLC  DFW      988          0         0   
457238     1267.0    1268.0     SEA  HNL     2677          0         0   
325724     1249.0    1179.0     EGE  MSP      788          0         0   
1906664    1225.0    1233.0     LAS  STL     1372          0         0   
1599350    1148.0    1146.0     SMF  MSP     1518          0         0   
315641     1142.0    1143.0     SHV  ORD      733          0         0   
1698976    1103.0    1100.0     DSM  DFW      624          0         0   
1879242    1103.0    1092.0     DTW  LGA      501          0         0   

         AvarageSpeed  Delay_WithinTime TaxiIn_and_TaxiOut  
521096         255.32             Delay               19.0  
1222506        507.81             Delay               27.0  
1400101        478.06             Delay               22.0  
457238         464.22             Delay               23.0  
325724         429.82             Delay               89.0  
1906664        560.00             Delay               25.0  
1599350        508.83             Delay               34.0  
315641         379.14             Delay                8.0  
1698976        430.34             Delay               36.0  
1879242        455.45             Delay               49.0  

[10 rows x 21 columns]

### Rutas más largas (seleccionadas las 10 primeras)

In [24]:
rutas_mas_larg = vuelos.drop_duplicates(subset = "Distance").sort_values(by = 'Distance', ascending = False).head(10)
rutas_mas_larg[['Origin', 'Dest', 'Distance']]

Origin Dest  Distance
1415598    EWR  HNL      4962
1682456    ATL  HNL      4502
1809932    ORD  HNL      4243
1147070    KOA  ORD      4213
447765     ORD  OGG      4184
1706278    MSP  HNL      3972
1414945    HNL  IAH      3904
691514     DFW  HNL      3784
1385569    OGG  DFW      3711
1174269    ATL  ANC      3417

### f) Nuevos datos que considere relevantes

#### Cantidad de vuelos por mes

In [25]:
vuelos_mes = vuelos['Month'].value_counts().reset_index()
vuelos_mes.columns=['Número del mes', 'Cantidad de vuelos']
vuelos_mes.sort_values('Cantidad de vuelos', ascending=False)


Número del mes  Cantidad de vuelos
0                3               20907
1                6               20695
2               12               20585
3                2               19378
4                7               18911
5                1               18764
6                8               16796
7                4               16087
8                5               15689
9               11               10914
10              10               10667
11               9                9788

Los meses de Diciembre, Junio y Marzo son en los que más vuelos se realizan, mientras que en el mes de Septiembre es el que menos.

#### Cantidad vuelos por día del mes

In [26]:

vuelos_Dia_Mes=vuelos['DayofMonth'].value_counts().reset_index()
vuelos_Dia_Mes.columns=['Día del Mes', 'Cantidad de vuelos']
vuelos_Dia_Mes.sort_values('Cantidad de vuelos', ascending=False)


Día del Mes  Cantidad de vuelos
0            22                7524
1            21                7244
2             1                6913
3             7                6891
4            15                6722
5            14                6689
6            27                6663
7            20                6646
8            18                6640
9            19                6636
10           11                6592
11           24                6577
12            2                6568
13            9                6560
14           17                6551
15           23                6532
16           10                6478
17            8                6451
18           26                6444
19           12                6341
20            4                6324
21            6                6324
22           13                6260
23           16                6254
24           28                6194
25            5                6168
26            3                6168
27           25                6077
28           29                5883
29           30                5881
30           31                3986

Llama la atención como precisamente los últimos días de cada mes (31, 30 y 29) es cuando se realizan menos vuelos, en cambio el día 1 está entre los tres primeros.

#### Cantidad vuelos por día de la semana

In [27]:
vuelos_Dia_Sem=vuelos['DayOfWeek'].value_counts().reset_index()
vuelos_Dia_Sem.columns=['Día de la semana', 'Cantidad de vuelos']
vuelos_Dia_Sem.sort_values('Cantidad de vuelos', ascending=False)


Día de la semana  Cantidad de vuelos
0                 5               33434
1                 1               29865
2                 4               29683
3                 7               29498
4                 3               26913
5                 2               26908
6                 6               22880

En cuanto a los vuelos que se efectúan por día de la semana, el primer y último día laborables de la semana son los días que más vuelos se ejecutan. Por su parte, el sábado es el que menos.

## Ejercicio 3

#### Exportar el nuevo dataframe con las nuevas columnas añadidas a Excel

In [ ]:
vuelos_colum_relev.to_excel('Vuelos_análisis.xlsx')